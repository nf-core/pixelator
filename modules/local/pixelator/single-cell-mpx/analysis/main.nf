process PIXELATOR_ANALYSIS {
    tag "${meta.id}"
    label 'process_medium'

    // TODO: Add conda back
    // conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(data)

    output:
    tuple val(meta), path("analysis/*dataset.pxl"),   emit: dataset
    tuple val(meta), path("analysis/*report.json"),   emit: report_json
    tuple val(meta), path("analysis/*.meta.json"),    emit: metadata
    tuple val(meta), path("analysis/*"),              emit: all_results
    tuple val(meta), path("*pixelator-analysis.log"), emit: log

    path "versions.yml",                              emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''

    """
    pixelator \\
        --cores ${task.cpus} \\
        --log-file ${prefix}.pixelator-analysis.log \\
        --verbose \\
        single-cell-mpx \\
        analysis \\
        --output . \\
        ${args} \\
        ${data}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir analysis
    touch "${prefix}.pixelator-analysis.log"
    touch "analysis/${prefix}.dataset.pxl"
    touch "analysis/${prefix}.report.json"
    touch "analysis/${prefix}.meta.json"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
