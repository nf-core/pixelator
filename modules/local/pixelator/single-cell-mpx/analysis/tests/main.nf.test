nextflow_process {

    name "Test Process PIXELATOR_ANALYSIS"
    script "../main.nf"
    process "PIXELATOR_ANALYSIS"
    tag "modules"
    tag "pixelator"
    tag "pixelator/mpx"
    tag "pixelator/single_cell_mpx_analysis"


    test("Test MPX analysis - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel:'human-sc-immunology-spatial-proteomics-1', technology:'mpx' ],
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/annotate/sample01_1k_pbmcs_scsp_v1_immunology1.annotate.dataset.pxl', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Test MPX analysis - SCSP v1 | Immunology-I") {
        when {
            params {
                save_all
            }
            process {
                """
                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel:'human-sc-immunology-spatial-proteomics-1', technology:'mpx' ],
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/annotate/sample01_1k_pbmcs_scsp_v1_immunology1.annotate.dataset.pxl', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out.report_json,
                        process.out.metadata
                    ).match()
                },
                {
                    with (process.out.dataset.get(0)) {
                        assert path(get(1)).zip.extract("adata.h5ad").exists()
                        assert path(get(1)).zip.extract("edgelist.parquet").exists()
                        assert snapshot(
                            path(get(1)).zip.extract("metadata.json")
                        ).match("pxl")
                    }
                },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator analysis")
                    }
                },
            )
        }
    }
}

