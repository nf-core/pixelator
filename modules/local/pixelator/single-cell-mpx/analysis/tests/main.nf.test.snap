{
    "Test MPX analysis - SCSP v1 | Immunology-I": {
        "content": [
            [
                [
                    {
                        "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                        "design": "D21",
                        "panel": "human-sc-immunology-spatial-proteomics-1",
                        "technology": "mpx"
                    },
                    "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,5db6473f8c444276e3b287231fc87481"
                ]
            ],
            [
                [
                    {
                        "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                        "design": "D21",
                        "panel": "human-sc-immunology-spatial-proteomics-1",
                        "technology": "mpx"
                    },
                    "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,f2597ab5ea83e5257f8c96330df16484"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:50:05.9291208"
    },
    "pxl": {
        "content": [
            "metadata.json:md5,57e6b62b023056bbe857829a5796bb7a"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-06T15:09:46.933037588"
    },
    "Test MPX analysis - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.dataset.pxl:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.dataset.pxl:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-analysis.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,21f7010150c44ce0f97dfe8a429047a1"
                ],
                "all_results": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.dataset.pxl:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "dataset": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.dataset.pxl:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-analysis.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "metadata": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "report_json": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,21f7010150c44ce0f97dfe8a429047a1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:48:02.493689902"
    }
}